src/gensearch.jl

```diff
@@ -0,0 +1,336 @@
+using Graphs
+
+## missing features:
+# - make functions callable with non-sets (e.g. X = 1, Y = 4) -> what is the best way to do this?
+
+const INIT = 0
+const LEFT = 1
+const RIGHT = 2
+
+## helper functions
+
+function inunion(w, sets...)
+    for S in sets
+        w in S && return true
+    end
+    return false
+end
+
+# maybe replace unused arguments by type like so:
+# function no_veto(::T, ::T, ::T, ::T)
+function no_veto(pe, ne, v, w)
+    return false
+end
+
+function no_outgoing(S)
+    return (pe, ne, v, w) -> v in S && ne == RIGHT 
+end
+
+function no_incoming(S)
+    return (pe, ne, v, w) -> v in S && ne == LEFT
+end
+
+
+function gensearch(g, S, pass)
+    visited = falses(3*nv(g))
+    function genvisit(g, v, pe)
+        visited[3*v - pe] = true
+        for ne in [LEFT, RIGHT]
+            ne == LEFT && (neighbors = inneighbors(g, v))
+            ne == RIGHT && (neighbors = outneighbors(g, v))
+            for w in neighbors
+                pass(pe, ne, v, w) && !visited[3*w - ne] && genvisit(g, w, ne)
+            end
+        end
+    end
+    foreach(s -> genvisit(g, s, INIT), S)
+    return Set([x for x in 1:nv(g) if visited[3*x-2] || visited[3*x-1] || visited[3*x]])
```

Comment:
Is this line called often?

Reply:
It is called once per call of gensearch.

gensearch returns the set of all visited vertices. Imho this is the most general/flexible way to do it.

In the context of d-separation, one call of gensearch with S = X (and the right pass function) will return the set of all vertices d-connected with X. To test whether Y is d-separated from X one could then check whether the intersection of X and Y is empty.


Alternatives would be:
(i) to return a boolean vector of visited vertices (less overhead, but also not so convenient to work with)
(ii) to write a second gensearch function which tests reachability (e.g., in  the d-separation example it would directly return a boolean). This could be used in cases where the whole set of visited vertices is not needed.

Reply:
Okay I think it’s fine

Reply:
```suggestion
    return Set(x for x in 1:nv(g) if visited[3*x-2] || visited[3*x-1] || visited[3*x])
```
would make sense.